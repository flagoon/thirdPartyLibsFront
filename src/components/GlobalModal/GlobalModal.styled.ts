import Grid from '@material-ui/core/Grid';
import Container from '@material-ui/core/Container';
import styled from 'styled-components';

export const Backdrop = styled(Container)`
  position: fixed;
  z-index: 100;
  background-color: rgba(0, 0, 0, 0.5);
  width: 100%;
  height: 100%;
  justify-content: center;
  margin: -16px -24px;
`;

export const ModalContainer = styled(Grid)`
  background-color: white;
  border: 5px solid black;
  margin: 5rem auto;
  width: 600px;
  height: 80%;
`;

export const ContainerTitle = styled(Grid)`
  border-bottom: 3px solid black;
`;

export const ContainerBody = styled(Grid)`
  max-height: 100%;
  flex-grow: 1;
  border-bottom: 3px solid black;
`;

export const ContainerFooter = styled(Grid)`
  background-color: white;
`;
