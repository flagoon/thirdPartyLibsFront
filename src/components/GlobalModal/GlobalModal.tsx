import React, { useRef, MouseEvent } from 'react';
import Button from '@material-ui/core/Button';
import {
  ModalContainer,
  Backdrop,
  ContainerBody,
  ContainerFooter,
  ContainerTitle,
} from './GlobalModal.styled';

interface Props {
  onModalClose: () => void;
}

const Modal = ({ onModalClose }: Props): JSX.Element => {
  const backdropRef = useRef<HTMLDivElement>(null);

  const onBackdropClick = (event: MouseEvent): void => {
    const isModalClicked =
      backdropRef &&
      backdropRef.current &&
      backdropRef.current.contains(event.target as Node);
    if (!isModalClicked) {
      onModalClose();
    }
  };

  return (
    <Backdrop maxWidth="xl" onClick={onBackdropClick} data-testid="backdrop">
      <ModalContainer
        direction="column"
        container
        spacing={2}
        ref={backdropRef}
      >
        <ContainerTitle item data-testid="containerTitle">
          <div>Title</div>
        </ContainerTitle>
        <ContainerBody item data-testid="containerBody">
          <h2>Content</h2>
        </ContainerBody>
        <ContainerFooter
          container
          item
          justify="flex-end"
          data-testid="containerFooter"
        >
          <Button component="label" data-testid="modalUploadButton">
            Upload File
            <input type="file" style={{ display: 'none' }} accept=".json" />
          </Button>
          <Button data-testid="modalAddRecordButton">Add record</Button>
          <Button onClick={onModalClose} data-testid="modalCancelButton">
            Cancel
          </Button>
        </ContainerFooter>
      </ModalContainer>
    </Backdrop>
  );
};

export default Modal;
