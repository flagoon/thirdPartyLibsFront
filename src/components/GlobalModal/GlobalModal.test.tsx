import React from 'react';
import GlobalModal from './GlobalModal';
import { render, fireEvent } from '@testing-library/react';

describe('GlobalModal', () => {
  it('should match snapshot', () => {
    const onModalClick = jest.fn();
    const { asFragment } = render(<GlobalModal onModalClose={onModalClick} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should call onBackdropClick function after clicking backdrop', () => {
    const onClickHandler = jest.fn();
    const { getByTestId } = render(
      <GlobalModal onModalClose={onClickHandler} />,
    );
    const backdrop = getByTestId('backdrop');
    fireEvent.click(backdrop);
    expect(onClickHandler).toHaveBeenCalledTimes(1);
  });
  it('should call onBackdropClick when clicking cancel button', () => {
    const onClickHandler = jest.fn();
    const { getByTestId } = render(
      <GlobalModal onModalClose={onClickHandler} />,
    );
    const cancelButton = getByTestId('modalCancelButton');
    fireEvent.click(cancelButton);
    expect(onClickHandler).toHaveBeenCalledTimes(1);
  });
  it('should not call onBackdropClick while clicking any other element', () => {
    const onClickHandler = jest.fn();
    const { getByTestId } = render(
      <GlobalModal onModalClose={onClickHandler} />,
    );
    fireEvent.click(getByTestId('containerTitle'));
    fireEvent.click(getByTestId('containerBody'));
    fireEvent.click(getByTestId('containerFooter'));
    fireEvent.click(getByTestId('modalUploadButton'));
    fireEvent.click(getByTestId('modalAddRecordButton'));
    expect(onClickHandler).toHaveBeenCalledTimes(0);
  });
});
