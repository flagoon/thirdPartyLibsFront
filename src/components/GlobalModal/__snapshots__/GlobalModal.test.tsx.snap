// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalModal should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root sc-AxjAm hTXoHK MuiContainer-maxWidthXl"
    data-testid="backdrop"
  >
    <div
      class="MuiGrid-root sc-AxirZ bGlnwV MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column"
    >
      <div
        class="MuiGrid-root sc-AxiKw iVCGtU MuiGrid-item"
        data-testid="containerTitle"
      >
        <div>
          Title
        </div>
      </div>
      <div
        class="MuiGrid-root sc-AxhCb gfYwBM MuiGrid-item"
        data-testid="containerBody"
      >
        <h2>
          Content
        </h2>
      </div>
      <div
        class="MuiGrid-root sc-AxhUy hAEkQL MuiGrid-container MuiGrid-item MuiGrid-justify-xs-flex-end"
        data-testid="containerFooter"
      >
        <label
          aria-disabled="false"
          class="MuiButtonBase-root MuiButton-root MuiButton-text"
          data-testid="modalUploadButton"
          role="button"
          tabindex="0"
        >
          <span
            class="MuiButton-label"
          >
            Upload File
            <input
              accept=".json"
              style="display: none;"
              type="file"
            />
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </label>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text"
          data-testid="modalAddRecordButton"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Add record
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text"
          data-testid="modalCancelButton"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Cancel
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
